code/modules/mob/living/simple_animal/hostile/pulsedemon.dm

```diff
@@ -0,0 +1,582 @@
+    if(!locate(/obj/structure/cable) in NewLoc || !locate(/obj/machinery/power) in NewLoc)
+        return
+    ..()
+    if(!is_under_tile() && prob(25))
+        spark(src,rand(2,4))
+    var/obj/machinery/power/new_power = locate(/obj/machinery/power) in NewLoc
+    if(new_power && !current_power)
+        current_power = new_power
+        current_cable = null
+        loc = new_power
+        playsound(src,'sound/weapons/electriczap.ogg',50, 1)
+        spark(src,rand(2,4))
+        if(istype(current_power,/obj/machinery/power/apc))
+            var/obj/machinery/power/apc/current_apc = current_power
+            if(current_apc.occupant)
+                to_chat(src,"<span class='warning'>Something else that isn't a pulse demon is already in here!</span>")
+                return
+            max_can_absorb = current_apc.cell.maxcharge
+            if(current_apc.pulsecompromised)
+                controlling_area = get_area(current_power)
+            else
+                hijackAPC(current_apc)
+        else if(istype(current_power,/obj/machinery/power/battery))
+            var/obj/machinery/power/battery/current_battery = current_power
+            to_chat(src,"<span class='notice'>You are now draining power from \the [current_power] and refilling charge.</span>")
+            max_can_absorb = current_battery.output
+    else
+        var/obj/structure/cable/new_cable = locate(/obj/structure/cable) in NewLoc
+        if(new_cable)
+            current_cable = new_cable
+            previous_net = current_net
+            current_net = current_cable.get_powernet()
+            current_power = null
+            current_robot = null
+            current_weapon = null
+            if(!isturf(loc))
+                loc = get_turf(NewLoc)
+            controlling_area = null
+            update_cableview()
+        else
+            current_cable = null
+            current_power = null
+            current_robot = null
+            current_weapon = null
+
+/mob/living/simple_animal/hostile/pulse_demon/to_bump(var/atom/obstacle) // Copied from how adminbus does it
+    if(!is_under_tile() && isliving(obstacle))
+        var/mob/living/L = obstacle
+        shockMob(L)
+    if(can_move && !locate(/obj/machinery/power) in get_turf(obstacle))
+        can_move = 0
+        forceMove(get_step(src,src.dir))
+        sleep(1)
+        can_move = 1
+    else
+        return ..()
+
+/obj/machinery/power/relaymove(mob/user, direction)
+    if(istype(user,/mob/living/simple_animal/hostile/pulse_demon))
+        var/turf/T = get_turf(src)
+        var/turf/T2 = get_step(T,direction)
+        if(locate(/obj/structure/cable) in T2)
+            playsound(src,'sound/weapons/electriczap.ogg',50, 1)
+            spark(src,rand(2,4))
+            user.forceMove(get_turf(src))
+
+/mob/living/simple_animal/hostile/pulse_demon/ClickOn(var/atom/A, var/params)
```

Comment:
surely most of this can be inherited instead of pasting it?

or done away with? like can pulse demons pick up guns??